jni/test/jni_array_test.dart

```diff
@@ -0,0 +1,232 @@
+    expect(array[0], true);
+    expect(array[1], true);
+    expect(array[2], true);
+    expect(() {
+      final _ = array[-1];
+    }, throwsRangeError);
+    expect(() {
+      array[-1] = false;
+    }, throwsRangeError);
+    expect(() {
+      array[3] = false;
+    }, throwsRangeError);
+    array.delete();
+  });
+  test("Java char array", () {
+    final array = JniCharTypeClass().newArray(3);
+    expect(array.length, 3);
+    array[0] = '1';
+    array[1] = '2';
+    array[2] = '3';
+    expect(array[0], '1');
+    expect(array[1], '2');
+    expect(array[2], '3');
+    array.setRange(0, 3, ['4', '5', '6', '7'], 1);
+    expect(array[0], '5');
+    expect(array[1], '6');
+    expect(array[2], '7');
+    expect(() {
+      final _ = array[-1];
+    }, throwsRangeError);
+    expect(() {
+      array[-1] = '4';
+    }, throwsRangeError);
+    expect(() {
+      array[3] = '4';
+    }, throwsRangeError);
+    array.delete();
+  });
+  test("Java byte array", () {
+    final array = JniByteTypeClass().newArray(3);
+    expect(array.length, 3);
+    array[0] = 1;
+    array[1] = 2;
+    array[2] = 3;
+    expect(array[0], 1);
+    expect(array[1], 2);
+    expect(array[2], 3);
+    array.setRange(0, 3, [4, 5, 6, 7], 1);
+    expect(array[0], 5);
+    expect(array[1], 6);
+    expect(array[2], 7);
+    expect(() {
+      final _ = array[-1];
+    }, throwsRangeError);
+    expect(() {
+      array[-1] = 4;
+    }, throwsRangeError);
+    expect(() {
+      array[3] = 4;
+    }, throwsRangeError);
+    array.delete();
+  });
+  test("Java short array", () {
+    final array = JniShortTypeClass().newArray(3);
+    expect(array.length, 3);
+    array[0] = 1;
+    array[1] = 2;
+    array[2] = 3;
+    expect(array[0], 1);
+    expect(array[1], 2);
+    expect(array[2], 3);
+    array.setRange(0, 3, [4, 5, 6, 7], 1);
+    expect(array[0], 5);
+    expect(array[1], 6);
+    expect(array[2], 7);
+    expect(() {
+      final _ = array[-1];
+    }, throwsRangeError);
+    expect(() {
+      array[-1] = 4;
+    }, throwsRangeError);
+    expect(() {
+      array[3] = 4;
+    }, throwsRangeError);
+    array.delete();
+  });
+  test("Java int array", () {
+    final array = JniIntTypeClass().newArray(3);
+    expect(array.length, 3);
+    array[0] = 1;
+    array[1] = 2;
+    array[2] = 3;
+    expect(array[0], 1);
+    expect(array[1], 2);
+    expect(array[2], 3);
+    array.setRange(0, 3, [4, 5, 6, 7], 1);
+    expect(array[0], 5);
+    expect(array[1], 6);
+    expect(array[2], 7);
+    expect(() {
+      final _ = array[-1];
+    }, throwsRangeError);
+    expect(() {
+      array[-1] = 4;
+    }, throwsRangeError);
+    expect(() {
+      array[3] = 4;
+    }, throwsRangeError);
+    array.delete();
+  });
+  test("Java float array", () {
+    final array = JniFloatTypeClass().newArray(3);
+    expect(array.length, 3);
+    array[0] = 0.5;
+    array[1] = 2;
+    array[2] = 3;
+    expect(array[0], 0.5);
```

Comment:
Shouldn't these expect a roughly equals? e.g. floats/doubles are rarely bitwise equal

Reply:
Hehe yeah, but in general with some other numbers int might not.

(Especially for float because we convert double in Dart to a float in C which then goes to jni.)

I think it's just best practise to expect approx instead of precise, so it doesn't suddenly break on a slightly different setup/hardware.

Reply:
Sure I'll change it to approx!